Deprecated NodeConfig.NodeIP in favor of NodeConfig.KubeletArguments["node-ip"]

pravisankar opened the pull request:
- Both NodeConfig.NodeIP and NodeConfig.KubeletArguments["node-ip"] optional params
  does the same job of setting Kubelet reported node address as given IP.
  When we have different node IP values for these params NodeConfig.KubeletArguments
  gets the precedence. Deprecating NodeConfig.NodeIP will avoid this confusion and
  it retains the same fuctionality.

- This also helps when we introduce PodTrafficNodeIP and MasterTrafficNodeIP for this
  card: https://trello.com/c/sCNKKYCz

Docs pr: https://github.com/openshift/openshift-docs/pull/6708

pravisankar commented:
@openshift/networking PTAL

liggitt commented:
we don't validate or require the kubelet flags. we do validate first-class config file fields. I don't anticipate us removing the first-class config field.

pravisankar commented:
@liggitt Let me add some context here. I added support for specifying desired node IP in the k8s upstream. To leverage that functionality in openshift, I added nodeIP param in node config (may be kubeletArguments was not there at that time or I may not be aware of this mechanism). Validation for this param is already done in the k8s. Currently this is just redundant and could lead to confusion when both nodeIP and kubeletArguments["node-ip"] are set.

pravisankar commented:
/retest

liggitt commented on pkg/cmd/server/api/serialization_test.go:
```diff
@@ -225,6 +226,19 @@ func fuzzInternalObject(t *testing.T, forVersion schema.GroupVersion, item runti
+				}
+
+				if _, ok := obj.KubeletArguments["node-ip"]; ok {
+					glog.Warning("nodeIP config is ignored in favor of kubeletArguments[\"node-ip\"]")
```
fuzz function is not the right place to print warnings...

liggitt commented on pkg/cmd/server/api/v1/conversions.go:
```diff
@@ -149,6 +151,18 @@ func SetDefaults_NodeConfig(obj *NodeConfig) {
+		}
+
+		if _, ok := obj.KubeletArguments["node-ip"]; ok {
+			glog.Warning("nodeIP config is ignored in favor of kubeletArguments[\"node-ip\"]")
```
don't print in defaulting/conversion... make it a warning or error in config validation

liggitt replied:
also, it's only a warning/error if they mismatch

liggitt commented on pkg/cmd/server/api/validation/node.go:
```diff
@@ -27,9 +27,6 @@ func ValidateInClusterNodeConfig(config *api.NodeConfig, fldPath *field.Path) Va
 	if len(config.NodeName) == 0 {
 		validationResults.AddErrors(field.Required(fldPath.Child("nodeName"), ""))
 	}
-	if len(config.NodeIP) > 0 {
```
restore validation, require it to be a valid IP, warn/error if it mismatches one specified in kubeletArgs[node-ip]

pravisankar commented:
@liggitt  updated, PTAL

pravisankar commented:
/retest

liggitt commented:
as kube config moves from flags to config files, we will have to transform our config into kube config objects. what is the benefit of moving from our structured config to flags now?

liggitt commented on pkg/cmd/server/api/validation/node.go:
```diff
@@ -27,8 +29,16 @@ func ValidateInClusterNodeConfig(config *api.NodeConfig, fldPath *field.Path) Va
+		if config.KubeletArguments != nil {
+			if ips, ok := config.KubeletArguments["node-ip"]; ok && (len(ips) > 0) {
+				if config.DeprecatedNodeIP != ips[0] {
+					glog.Warning("nodeIP config is ignored in favor of kubeletArguments[\"node-ip\"]")
```
why not an error?

liggitt commented on pkg/cmd/server/start/bootstrap_node.go:
```diff
@@ -139,7 +139,9 @@ func overrideNodeConfigForBootstrap(nodeConfig *configapi.NodeConfig, bootstrapK
 	nodeConfig.NodeName = nodeutil.GetHostname(nodeConfig.NodeName)
 	if nodeConfig.DNSIP == "0.0.0.0" {
-		nodeConfig.DNSIP = nodeConfig.NodeIP
+		if ips, ok := nodeConfig.KubeletArguments["node-ip"]; ok {
```
check len(ips) > 0

liggitt commented on pkg/cmd/server/kubernetes/network/sdn_linux.go:
```diff
@@ -38,10 +38,17 @@ func NewSDNInterfaces(options configapi.NodeConfig, networkClient networkclient.
 
+	nodeIP := ""
+	if options.KubeletArguments != nil {
+		if ips, ok := options.KubeletArguments["node-ip"]; ok {
```
len(ips) > 0

liggitt commented on pkg/cmd/server/api/validation/node.go:
```diff
@@ -27,8 +29,16 @@ func ValidateInClusterNodeConfig(config *api.NodeConfig, fldPath *field.Path) Va
+		if config.KubeletArguments != nil {
+			if ips, ok := config.KubeletArguments["node-ip"]; ok && (len(ips) > 0) {
+				if config.DeprecatedNodeIP != ips[0] {
+					glog.Errorf("nodeIP config is ignored in favor of kubeletArguments[\"node-ip\"]")
```
not just a log error, add an error to the validationResults

pravisankar commented:
"what is the benefit of moving from our structured config to flags now?"
I mentioned this on IRC, nodeIP is *not* openshift config, it is actually kubelet config and it incorrectly ended in our structured openshift config. We have to treat this flag just like any other kubelet args (fail-swap-on, cgroups-per-qos, etc.)

pravisankar commented:
@liggitt changed to validation error as suggested, PTAL

liggitt commented:
/lgtm

pravisankar commented:
@liggitt rebased, ptal

liggitt commented:
/lgtm

pravisankar commented:
/retest

dcbw commented:
"Unschedulable no nodes available to schedule pods"

GCE suckage perhaps?

/test extended_conformance_gce

kargakis commented:
/retest

smarterclayton commented:
This might be breaking the control plane in bootstrap mode.  You've failed a lot of GCP runs.

/retest

smarterclayton commented:
This has *never* passed GCP.  That's pretty surprising.

liggitt commented:
This is the failing task:

FAILED - RETRYING: Verify that the web console is running (1 retries left).
fatal: [ci-prtest-5a37c28-15047-ig-m-4zbx]: FAILED! => {"attempts": 120, "changed": false, "cmd": ["curl", "-k", "https://webconsole.openshift-web-console.svc/healthz"], "delta": "0:00:00.019781", "end": "2018-01-26 00:41:38.447063", "msg": "non-zero return code", "rc": 6, "start": "2018-01-26 00:41:38.427282", "stderr": "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dload  Upload   Total   Spent    Left  Speed\n\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: webconsole.openshift-web-console.svc; Name or service not known", "stderr_lines": ["  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current", "                                 Dload  Upload   Total   Spent    Left  Speed", "", "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: webconsole.openshift-web-console.svc; Name or service not known"], "stdout": "", "stdout_lines": []}

pravisankar commented:
/retest

pravisankar commented:
/retest
